### ISSUE TITLE
Hashie gem not working

### ISSUE THREAD

**Issue by brauliobo** [oracle/truffleruby#1333]
MRI:
```ruby
2.4.2 :002 > require 'hashie'
 => true
2.4.2 :003 > h=Hashie::Mash.new({a: {}, b: {}})
 => #<Hashie::Mash a=#<Hashie::Mash> b=#<Hashie::Mash>>
2.4.2 :004 > h.a.blah = 5
 => 5 
```

TruffleRuby:
```ruby
irb(main):003:0> require 'hashie'
=> true
irb(main):006:0> h=Hashie::Mash.new({a: {}, b: {}})
=> #<Hashie::Mash>
irb(main):007:0> h.a.blah = 5
NoMethodError: undefined method `blah=' for nil:NilClass
        from (irb):7:in `method_missing'
        from (irb):7:in `<top (required)>'
        from /home/braulio/Projects/graalvm-1.0.0-rc1/jre/languages/ruby/bin/irb:29:in `<main>'
```

**Comment by eregon** [comment]
Hello,
Thank you for the report.

I can reproduce with hashie 3.5.7:
```bash
$ gem i hashie
Fetching: hashie-3.5.7.gem (100%)
Successfully installed hashie-3.5.7
1 gem installed
$ ruby -rhashie -e 'h=Hashie::Mash.new({a: {}, b: {}}); h.a.blah = 5; p h'
#<Hashie::Mash a=#<Hashie::Mash blah=5> b=#<Hashie::Mash>>

$ jt ruby -S gem i hashie                                                 
Fetching: hashie-3.5.7.gem (100%)
Successfully installed hashie-3.5.7
1 gem installed

$ jt ruby -rhashie -e 'h=Hashie::Mash.new({a: {}, b: {}}); h.a.blah = 5; p h'
-e:1:in `method_missing': undefined method `blah=' for nil:NilClass (NoMethodError)
	from -e:1:in `<main>'

$ jt ruby -rhashie -e 'h=Hashie::Mash.new({a: {}, b: {}}); p h.a; p h'       
nil
#<Hashie::Mash>
```

So it looks like `Hashie::Mash.new({a: {}, b: {}})` somehow ignores the arguments on TruffleRuby.

**Comment by eregon** [comment]
I found the problem for this example:
In Hashie::Mash there is:
```ruby
    def initialize(source_hash = nil, default = nil, &blk)
      deep_update(source_hash) if source_hash
      default ? super(default) : super(&blk)
    end
```
The last line calls Hash#initialize, and that for us reset the Hash storage to empty.
It seems MRI doesn't touch the Hash on initialize with no arguments, so we'll add a spec for that and fix the behavior.

**Comment by eregon** [comment]
Fix in #1334.
The semantics of MRI are that Hash#initialize changes the default value and proc, but nothing else.

**Comment by brauliobo** [comment]
thank you for the quick fix
